### PR TITLE
Only allow expanding generics to form intersection constraints a limited number of times

### LINKED ISSUE #46900: tsc 4.5 hangs with unist-util-visit package (works fine with 4.4.4)
# Bug Report

the following example makes `tsc` hang indefinitely with version 4.5, but compiles fine with 4.4.4:

example: https://github.com/stefanprobst/issue-tsc

also see the discussion here: https://github.com/unifiedjs/unified/discussions/175 

### 🔎 Search Terms

hangs

### 🕗 Version & Regression Information

- This is a crash: `tsc` hangs, but does not give an error message
- This changed between versions 4.4.4 and  4.5.2 (same behavior also in current `@next` version)

### ⏯ Playground Link

link to a reproduction repo above

### 🙁 Actual behavior

`tsc` hangs indefinitely without error message

### 🙂 Expected behavior

`tsc` finishing compilation, or giving indication what's wrong

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #46900

This prevents us from (slowly) infinitely expanding our constraints as we're comparing intersections of generics to unions.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -19387,13 +19400,7 @@ namespace ts {
...
                         // conditionals aren't related to one another via distributive constraint as it is much too inaccurate and allows way
                         // more assignments than are desirable (since it maps the source check type to its constraint, it loses information)
-                        const distributiveConstraint = hasNonCircularBaseConstraint(source) ? getConstraintOfDistributiveConditionalType(source as ConditionalType) : undefined;
-                        if (distributiveConstraint) {
-                            if (result = isRelatedTo(distributiveConstraint, target, RecursionFlags.Source, reportErrors)) {
-                                resetErrorInfo(saveErrorInfo);
-                                return result;
-                            }
-                        }
+                        doDistributiveCheck = true;
```
Comment (weswigham) [inline src/compiler/checker.ts:19403]: This just move checking of the distributive constraint to _after_ checking the normal constraint, which is needed to remove an instantiation depth error from the example repo in

**Comment by weswigham** [comment]
I don't yet have a truly minimal repro from #46900 (just the submission that uses ~5 packages of types), because the types involved are quite complicated, however I'm going to endeavor to create one now that I have a fix that I know works ready.

**Comment by ahejlsberg** [comment]
I'm not sure we need this PR as the issue appears to already be fixed by #47738. That PR moves logic from the `isRelatedTo` function (the uncached side of relationship checking) to the `structuredTypeRelatedTo` function (the cached side) where it should have been in the first place. Apart from the caching benefits this makes the logic subject to recursion depth checking which cuts off the infinite intersection constraint generation in #46900.

I think we should just merge #47738 for 4.6. I took Daniel's comment [here](https://github.com/microsoft/TypeScript/pull/47738#pullrequestreview-879332099) to mean hold until after 4.6, but I do think the changes are safe.

**Comment by weswigham** [comment]
Superseded by #47738

**Pull request closed without merging**